src/url_search_params.rs

```diff
@@ -0,0 +1,483 @@
+impl<'a> Iterator for UrlSearchParamsValueIterator<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let has_next = unsafe { ffi::ada_search_params_values_iter_has_next(self.iterator) };
+        if has_next {
+            let string = unsafe { ffi::ada_search_params_values_iter_next(self.iterator) };
+            Some(string.as_str())
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a> UrlSearchParamsValueIterator<'a> {
+    fn new(
+        iterator: *mut ffi::ada_url_search_params_values_iter,
+    ) -> UrlSearchParamsValueIterator<'a> {
+        UrlSearchParamsValueIterator {
+            iterator,
+            _phantom: core::marker::PhantomData,
+        }
+    }
+}
+
+pub struct UrlSearchParamsEntry<'a> {
+    strings: *mut ffi::ada_strings,
+    size: usize,
+    _phantom: core::marker::PhantomData<&'a str>,
+}
+
+impl<'a> UrlSearchParamsEntry<'a> {
+    fn new(strings: *mut ffi::ada_strings, size: usize) -> UrlSearchParamsEntry<'a> {
+        UrlSearchParamsEntry {
+            strings,
+            size,
+            _phantom: core::marker::PhantomData,
+        }
+    }
+
+    /// Returns whether the key value pair is empty or not
+    ///
+    /// ```
+    /// use ada_url::UrlSearchParams;
+    /// let params = UrlSearchParams::parse("a=1&b=2")
+    ///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+    /// let pairs = params.get_all("a");
+    /// assert_eq!(pairs.is_empty(), false);
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.size == 0
+    }
+
+    /// Returns the size of the key value pairs
+    ///
+    /// ```
+    /// use ada_url::UrlSearchParams;
+    /// let params = UrlSearchParams::parse("a=1&b=2")
+    ///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+    /// let pairs = params.get_all("a");
+    /// assert_eq!(pairs.len(), 1);
+    /// ```
+    pub fn len(&self) -> usize {
+        self.size
+    }
+
+    /// Get an entry by index
+    ///
+    /// ```
+    /// use ada_url::UrlSearchParams;
+    /// let params = UrlSearchParams::parse("a=1&a=2")
+    ///     .expect("This is a valid UrlSearchParams. Should have parsed it.");
+    /// let pairs = params.get_all("a");
+    /// assert_eq!(pairs.len(), 2);
+    /// assert_eq!(pairs.get(0), Some("1"));
+    /// assert_eq!(pairs.get(1), Some("2"));
+    /// assert_eq!(pairs.get(2), None);
+    /// assert_eq!(pairs.get(55), None);
+    /// ```
+    pub fn get(&self, index: usize) -> Option<&str> {
+        if self.size == 0 || index > self.size - 1 {
+            return None;
+        }
+
+        unsafe {
+            let string = ffi::ada_strings_get(self.strings, index);
+            Some(string.as_str())
+        }
+    }
+}
+
+impl Drop for UrlSearchParamsEntry<'_> {
+    fn drop(&mut self) {
+        unsafe { ffi::ada_free_strings(self.strings) }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<'a> From<UrlSearchParamsEntry<'a>> for Vec<&'a str> {
```

Comment:
Just looked up the conventions for this (https://doc.rust-lang.org/std/convert/trait.Into.html):

>One should avoid implementing [Into](https://doc.rust-lang.org/std/convert/trait.Into.html) and implement [From](https://doc.rust-lang.org/std/convert/trait.From.html) instead. Implementing [From](https://doc.rust-lang.org/std/convert/trait.From.html) automatically provides one with an implementation of [Into](https://doc.rust-lang.org/std/convert/trait.Into.html) thanks to the blanket implementation in the standard library.

It still feels a bit weird to see a `From` for a foreign type. But I guess its fine :-)

Reply:
I agree. I implemented Into in the first place and clippy recommended me to rewrite.